File: Notepad++/02_before/190219quicksort.cpp
#include <iostream>
#include <vector>
using namespace std;

int partition(vector<int>& , int, int);

void quickSort(vector<int>& a, int f, int l)
{
	int m;
	if( f < l)
	{
		m = partition(a, f, l);
		quickSort(a, f, m-1);
		quickSort(a, m+1, l);
	}
}

int partition(vector<int>& a, int f, int l)
{
	int key = a[l];
	int i = f - 1;
	int temp;
	for(int j = f; j < l; j++)
	{
		if(key >= a[j])
		{
			i += 1;
			temp = a[j];
			a[j] = a[i];
			a[i] = temp;
		}
	}
	temp = a[i+1];
	a[i+1] = a[l];
	a[l] = temp;
	
	return i+1;
}


int main()
{
	vector<int> a = {1, 2, 0, 3, 2, 1, 4, 5, 3, 8, 1};
	quickSort(a, 0, 10);
	for(int i = 0; i < 11; i++)
	{
		cout << a[i] << " ";
	}
	cout << endl;
	return 0;
}

File: Notepad++/02_before/190219quicksort.py
import numpy as np

def quickSort(x, a, b):
	if a < b:
		m = partition(x, a, b)
		quickSort(x, a, m-1)
		quickSort(x, m+1, b)
		
def partition(x, a, b):
	key = x[b]
	i = a - 1
	for j in range(a, b):
		if x[j] <= key:
			i += 1
			x[i], x[j] = x[j], x[i]
	x[i+1], x[b] = x[b], x[i+1]
	return i+1

test = np.random.randint(100, size=5)
print(test)

quickSort(test, 0, 4)
print(test)